src/tls.rs

```diff
@@ -1,6 +1,234 @@
+        }
+
+        let mut alpn_protos: Vec<*const i8>;
+        if let Some(protos) = cfg.alpn_protos {
+            alpn_protos = protos
+                .iter()
+                .map(|p| p.as_ptr())
+                .chain(std::iter::once(std::ptr::null()))
+                .collect();
+            rcfg.alpn_protos = alpn_protos.as_mut_ptr();
+        }
+
+        rcfg.non_block = cfg.non_block;
+        rcfg.use_secure_element = cfg.use_secure_element;
+        rcfg.timeout_ms = cfg.timeout_ms as i32;
+        rcfg.use_global_ca_store = cfg.use_global_ca_store;
+        rcfg.common_name = cfg.common_name.as_ptr();
+        rcfg.skip_common_name = cfg.skip_common_name;
+
+        let mut raw_kac: esp_idf_sys::tls_keep_alive_cfg;
+        if let Some(kac) = cfg.keep_alive_cfg {
+            raw_kac = esp_idf_sys::tls_keep_alive_cfg {
+                keep_alive_enable: kac.enable,
+                keep_alive_idle: kac.idle.as_secs() as i32,
+                keep_alive_interval: kac.interval.as_secs() as i32,
+                keep_alive_count: kac.count as i32,
+            };
+            rcfg.keep_alive_cfg = &mut raw_kac as *mut _;
+        }
+
+        let mut raw_psk: esp_idf_sys::psk_key_hint;
+        if let Some(psk) = cfg.psk_hint_key {
+            raw_psk = esp_idf_sys::psk_key_hint {
+                key: psk.key.as_ptr(),
+                key_size: psk.key.len(),
+                hint: psk.hint.as_ptr(),
+            };
+            rcfg.psk_hint_key = &mut raw_psk as *mut _;
+        }
+
+        rcfg.is_plain_tcp = cfg.is_plain_tcp;
+        rcfg.if_name = std::ptr::null_mut();
+
+        let ret = unsafe {
+            esp_idf_sys::esp_tls_conn_new_sync(
+                host.as_bytes().as_ptr() as *const i8,
+                host.len() as i32,
+                port,
+                &rcfg,
+                tls,
+            )
+        };
+
+        if ret == 1 {
+            Ok(EspTls { raw: tls })
+        } else {
+            unsafe {
+                esp_idf_sys::esp_tls_conn_destroy(tls);
+            }
+
+            Err(Error::ConnectionNotEstablished)
+        }
+    }
+
+    pub fn close(self) {}
+}
+
+impl Read for EspTls {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        // cannot call esp_tls_conn_read bc it's inline
+        //let ret = unsafe { esp_idf_sys::esp_tls_conn_read(self.raw, buf.as_mut_ptr(), buf.len()) };
+        let esp_tls = unsafe { std::ptr::read_unaligned(self.raw) };
+        let read_func = esp_tls.read.unwrap();
+        let ret = unsafe { read_func(self.raw, buf.as_mut_ptr() as *mut i8, buf.len()) } as i32;
+        // ESP docs treat 0 as error, but in Rust it's common to return 0 from `Read::read` to indicate eof
+        if ret >= 0 {
+            Ok(ret as usize)
+        } else {
+            Err(io::Error::new(io::ErrorKind::Other, Error::ReadError(ret)))
+        }
+    }
+}
+
+impl Write for EspTls {
```

Comment:
* Implement `embedded_svc::io::Write` instead 
* Move `write` as `pub fn write` on `EspTls` and then delegate to this method from here

Additional topic: is `EspTls` supporting parallel `read`/`write`? I would assume so. This means it should be possible to `EspTls::split(&mut self) -> (EspTlsRead<_>, EspTlsWrite<'_>)`, so that you can handle read separately from write.